pareto/strategic_water_management/strategic_produced_water_optimization.py

```diff
@@ -215,6 +218,489 @@ def get_strategic_model_unit_container():
+def get_piping_arc_list():
+    return [
+        "PCA",
+        "PNA",
+        "PPA",
+        "CNA",
+        "CCA",
+        "NNA",
+        "NCA",
+        "NKA",
+        "NSA",
+        "NRA",
+        "NOA",
+        "FCA",
+        "RNA",
+        "RCA",
+        "RKA",
+        "RSA",
+        "SNA",
+        "SCA",
+        "SKA",
+        "SRA",
+        "SOA",
+        "ROA",
+    ]
+
+
+# TODO: Move this function to a utility
+def _check_optional_data(
+    df_sets,
+    df_parameters,
+    optional_tab_name,
+    required_sets_with_option,
+    required_parameters_with_option,
+    requires_at_least_one=[],
+):
+    # create set object for df_sets and df_parameters for simpler item checking
+    _df_sets_set = set(df_sets)
+    _df_parameters_set = set(df_parameters.keys())
+    if optional_tab_name in _df_sets_set:
+        _missing_parameters = set(required_parameters_with_option) - _df_parameters_set
+        _missing_sets = set(required_sets_with_option) - _df_sets_set
+        if len(_missing_parameters) > 0:
+            for param in _missing_parameters:
+                df_parameters[param] = {}
+            warning_message = "Warning: {} are given, but {} Parameter data is missing. The following parameters have been assumed empty:".format(
+                optional_tab_name, optional_tab_name
+            ) + str(
+                _missing_parameters
+            )
+            warnings.warn(warning_message)
+        if len(_missing_sets) > 0:
+            for s in _missing_sets:
+                df_sets[s] = {}
+            warning_message = "Warning: {} are given, but other {} Set data is missing. The following set has been assumed to be empty:".format(
+                optional_tab_name, optional_tab_name
+            ) + str(
+                _missing_sets
+            )
+            warnings.warn(warning_message)
+        if len(requires_at_least_one) > 0:
+            for requires_list in requires_at_least_one:
+                _included_params = set(requires_list) & _df_parameters_set
+                if len(_included_params) < 1:
+                    warning_message = "Warning: {} are given, but some piping and trucking arcs are missing. At least one of the following arcs are required, missing sets have been assumed to be empty:".format(
+                        optional_tab_name
+                    ) + str(
+                        requires_list
+                    )
+                    warnings.warn(warning_message)
+
+    # If there is specific parameter data for optional tab, the optional tab name is also required
+    if (
+        len(required_parameters_with_option & _df_parameters_set) > 0
+    ) and optional_tab_name not in df_sets:
+        raise MissingDataError(
+            'Essential data is incomplete. Parameters for {} are given, but the "{}" Set is not given. Please add and complete the following tab(s): {}'.format(
+                optional_tab_name, optional_tab_name, optional_tab_name
```

Comment:
I'm not sure if this is correct - why is `optimal_tab_name` being printed at all three of the specified spots in this formatted string?

Reply:
I updated it slightly. This warning displays if the optional_set_name is missing from the input and required data dependent on it is included (e.g. "SWDSites" is missing, but "DisposalOperationalCost" is included). The warning will look like:

'Essential data is incomplete. Parameter data for {SWDSites} is given, but the "{SWDSites}" Set is missing. Please add and complete the following tab(s): {SWDSites}, or remove the following Parameters:{DisposalOperationalCost}'